File: Python/non-overlapping-intervals.py
# Time:  O(nlogn)
# Space: O(1)

class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        if not intervals or len(intervals) == 1:
            return 0
        
        intervals.sort()
        rms = 0
        pre_end = intervals[0][1]
        for start, end in intervals[1:]:
            if start < pre_end:
                rms += 1
                pre_end = min(end, pre_end)
            else:
                pre_end = end
        return rms

File: C++/non-overlapping-intervals.cpp
// Time:  O(nlogn)
// Space: O(1)

/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
    int eraseOverlapIntervals(vector<vector<int>>& intervals) {

	if (intervals.empty()) return 0;

	sort(intervals.begin(), intervals.end());
	int cnt = -1;
	int prev = INT_MAX;
	for (vector<int>& interval : intervals) {
		if (interval[0] < prev) {
			++cnt;
			prev = min(prev, interval[1]);
			continue;
		}
		prev = interval[1];
	}
	return cnt;
}
};
